src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Security/UseAutoValidateAntiforgeryToken.cs

```diff
@@ -195,21 +195,47 @@ public override void Initialize(AnalysisContext context)
 
                                 return;
                             }
-                            else if (potentialAntiForgeryFilter.AllInterfaces.Contains(iAsyncAuthorizationFilterTypeSymbol) ||
-                                potentialAntiForgeryFilter.AllInterfaces.Contains(iAuthorizationFilterTypeSymbol))
+                            else if (potentialAntiForgeryFilter.AllInterfaces.Contains(iAsyncAuthorizationFilterTypeSymbol))
                             {
-                                onAuthorizationAsyncMethodSymbols.TryAdd(
+                                // ASP.NET Core MVC seems to prioritize asynchronous over synchronous methods.
+                                // https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Mvc/Mvc.Core/src/Infrastructure/ResourceInvoker.cs#L311
+                                IMethodSymbol? onAuthorizationAsyncMethodSymbol =
                                     potentialAntiForgeryFilter
-                                    .GetMembers()
-                                    .OfType<IMethodSymbol>()
-                                    .FirstOrDefault(
-                                        s => (s.Name == "OnAuthorizationAsync" &&
-                                            s.ReturnType.Equals(taskTypeSymbol) ||
-                                            s.Name == "OnAuthorization" &&
-                                            s.ReturnsVoid) &&
-                                            s.Parameters.Length == 1 &&
-                                            s.Parameters[0].Type.Equals(authorizationFilterContextTypeSymbol)),
-                                    placeholder);
+                                        .GetBaseTypesAndThis()
+                                        .SelectMany(s => s.GetMembers())
+                                        .OfType<IMethodSymbol>()
+                                        .FirstOrDefault(
```

Comment:
In case of multiple matches, should we take all of them instead of the first only?

Reply:
Yes, we should take the first method, starting at the derived class and ending at the base class. `Test_GlobalAntiForgeryFilter_DerivedFilterOverrides_Diagnostic` is an example of that scenario.